packages/monaco/src/browser/monaco-text-model-service.ts

```diff
@@ -57,14 +67,48 @@ export class MonacoTextModelService implements monaco.editor.ITextModelService {
     }
 
     protected async loadModel(uri: URI): Promise<MonacoEditorModel> {
-        await this.editorPreferences.ready;
-        const resource = await this.resourceProvider(uri);
-        const model = await (new MonacoEditorModel(resource, this.m2p, this.p2m, { encoding: this.editorPreferences.get('files.encoding') }).load());
-        this.updateModel(model);
-        model.textEditorModel.onDidChangeLanguage(() => this.updateModel(model));
-        const disposable = this.editorPreferences.onPreferenceChanged(change => this.updateModel(model, change));
-        model.onDispose(() => disposable.dispose());
-        return model;
+        try {
+            await this.editorPreferences.ready;
+            const resource = await this.resourceProvider(uri);
+            const model = await (new MonacoEditorModel(resource, this.m2p, this.p2m, { encoding: this.editorPreferences.get('files.encoding') }).load());
+            this.updateModel(model);
+            model.textEditorModel.onDidChangeLanguage(() => this.updateModel(model));
+            const disposable = this.editorPreferences.onPreferenceChanged(change => this.updateModel(model, change));
+            model.onDispose(() => disposable.dispose());
+            return model;
+        } catch (error) {
+            const msg = 'Unable to open "' + uri + '": ' + error;
+            this.messages.showMessage({ type: MessageType.Error, text: msg, actions: ['CREATE'] })
+                .then(res => {
+                    if (res === 'CREATE') {
+                        this.getOrCreateDirectory(uri.parent).then(parent => {
+                            if (parent) {
+                                const parentUri = new URI(parent.uri);
+                                const name = uri.path.name;
+                                const ext = uri.path.ext;
+                                const fileUri = parentUri.resolve(name) +
+                                    (ext && ext !== null ? '.' && ext : '');
+
+                                this.fileSystem.createFile(fileUri.toString()).then(() => {
+                                    this.commands.executeCommand('vscode.open', fileUri);
+                                });
+                            }
+                        });
+                    }
+                });
+            return Promise.reject(error);
+        }
+    }
+
+    protected async getOrCreateDirectory(uri: URI): Promise<FileStat | undefined> {
+        const stat = await this.fileSystem.getFileStat(uri.toString());
```

Comment:
monaco model should not rely on filesystem directly, it can work with any kind of resources, e.g. git, preview and so on